pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/LookupTransformFunction.java

```diff
@@ -0,0 +1,332 @@
+  private String _dimTableName;
+  private String _dimColumnName;
+  private final List<String> _joinKeys = new ArrayList<>();
+  private final List<FieldSpec> _joinValueFieldSpecs = new ArrayList<>();
+  private final List<TransformFunction> _joinValueFunctions = new ArrayList<>();
+
+  private DimensionTableDataManager _dataManager;
+  private FieldSpec _lookupColumnFieldSpec;
+
+  @Override
+  public String getName() {
+    return FUNCTION_NAME;
+  }
+
+  @Override
+  public void init(List<TransformFunction> arguments, Map<String, DataSource> dataSourceMap) {
+    // Check that there are correct number of arguments
+    Preconditions
+        .checkArgument(arguments.size() >= 4,
+            "At least 4 arguments are required for LOOKUP transform function: " +
+            "LOOKUP(TableName, ColumnName, JoinKey, JoinValue [, JoinKey2, JoinValue2 ...])");
+    Preconditions
+        .checkArgument(arguments.size() % 2 == 0, "Should have the same number of JoinKey and JoinValue arguments");
+
+    TransformFunction dimTableNameFunction = arguments.get(0);
+    Preconditions.checkArgument(dimTableNameFunction instanceof LiteralTransformFunction,
+        "First argument must be a literal(string) representing the dimension table name");
+    _dimTableName = TableNameBuilder.OFFLINE.tableNameWithType(
+        ((LiteralTransformFunction) dimTableNameFunction).getLiteral());
+
+    TransformFunction dimColumnFunction = arguments.get(1);
+    Preconditions.checkArgument(dimColumnFunction instanceof LiteralTransformFunction,
+        "Second argument must be a literal(string) representing the column name from dimension table to lookup");
+    _dimColumnName = ((LiteralTransformFunction) dimColumnFunction).getLiteral();
+
+    List<TransformFunction> joinArguments = arguments.subList(2, arguments.size());
+    int numJoinArguments = joinArguments.size();
+    for (int i = 0; i < numJoinArguments / 2; i++) {
+      TransformFunction dimJoinKeyFunction = joinArguments.get((i * 2));
+      Preconditions.checkArgument(dimJoinKeyFunction instanceof LiteralTransformFunction,
+          "JoinKey argument must be a literal(string) representing the primary key for the dimension table");
+      _joinKeys.add(((LiteralTransformFunction) dimJoinKeyFunction).getLiteral());
+
+      TransformFunction factJoinValueFunction = joinArguments.get((i * 2) + 1);
+      TransformResultMetadata factJoinValueFunctionResultMetadata = factJoinValueFunction.getResultMetadata();
+      Preconditions.checkArgument(factJoinValueFunctionResultMetadata.isSingleValue(),
+          "JoinValue argument must be a single value expression");
+      _joinValueFunctions.add(factJoinValueFunction);
+    }
+
+    // Validate lookup table and relevant columns
+    _dataManager = DimensionTableDataManager.getInstanceByTableName(_dimTableName);
+    Preconditions.checkArgument(_dataManager != null,
+            String.format("Dimension table does not exist: %s", _dimTableName));
```

Comment:
No need to format string as `Preconditions` supports `%s`, same for other places

Reply:
Oh I didn't know about that, thanks for the tip. Updated.